service/agentconf.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// This file has routines which can be used for agent specific functionalites related to service files:
+//	- finding all agents in the machine
+//	- create conf file using the machine details
+// 	- write systemd service file and setting links
+// 	- copy all tools and related to agents and setup the links
+// 	- start all the agents
+// These routines can be used by any tools/cmds trying to implement the above functionality as part of the process, eg. juju-updateseries command.
+
+package service
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/arch"
+	"github.com/juju/utils/fs"
+	"github.com/juju/utils/shell"
+	"github.com/juju/utils/symlink"
+	"github.com/juju/version"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/agent/tools"
+	"github.com/juju/juju/juju/paths"
+	"github.com/juju/juju/service/common"
+	"github.com/juju/juju/service/systemd"
+)
+
+// FindAgents find all the agents available in the machine.
+func FindAgents(dataDir string) (machineAgent string, unitAgents []string, errAgentNames []string, err error) {
+
+	agentDir := filepath.Join(dataDir, "agents")
+	dir, err := os.Open(agentDir)
+	if err != nil {
+		return "", nil, nil, errors.Annotate(err, "opening agents dir")
+	}
+	defer dir.Close()
+
+	entries, err := dir.Readdir(-1)
+	if err != nil {
+		return "", nil, nil, errors.Annotate(err, "reading agents dir")
+	}
+	for _, info := range entries {
+		name := info.Name()
+		tag, err := names.ParseTag(name)
+		if err != nil {
+			continue
+		}
+		switch tag.Kind() {
+		case names.MachineTagKind:
+			machineAgent = name
+		case names.UnitTagKind:
+			unitAgents = append(unitAgents, name)
+		default:
+			errAgentNames = append(errAgentNames, name)
+			logger.Infof("%s is not of type Machine nor Unit, ignoring", name)
+		}
+	}
+	return machineAgent, unitAgents, errAgentNames, nil
+}
+
+// WriteSystemdAgents creates systemd files and create symlinks for the list of machine and units passed in the standard filepath '/var/lib/juju' during the upgrade process.
+func WriteSystemdAgents(machineAgent string, unitAgents []string, dataDir string, symLinkSystemdDir string, symLinkSystemdMultiUserDir string, series string) (errSysServiceNames []string, errSymServiceNames []string, errAgentNames []string, lastError error) {
```

Comment:
Same here - these should just be local vars rather than named return values.

I think `errSysServiceNames` and `errSymServiceNames` are misnamed - the values that get added to those slices aren't for errors are they? They're also a bit too similar to each other, only one character different! Could you give them names that make it clearer what they contain?